include/mbedtls/ccm_internal.h

```diff
@@ -0,0 +1,128 @@
+
+/**
+* \brief           default CCM initialization (encryption and decryption)
+*
+* \param ctx       default CCM context to be initialized
+* \param cipher    cipher to use (a 128-bit block cipher)
+* \param key       encryption key
+* \param keybits   key size in bits (must be acceptable by the cipher)
+*
+* \return          0 if successful, or a cipher specific error code
+*/
+int mbedtls_ccm_setkey_default( mbedtls_ccm_context_default *ctx,
+                                mbedtls_cipher_id_t cipher,
+                                const unsigned char *key,
+                                unsigned int keybits );
+
+/**
+* \brief           Free a CCM context and underlying cipher sub-context
+*
+* \param ctx       CCM context to free
+*/
+void mbedtls_ccm_free_default( mbedtls_ccm_context_default *ctx );
+
+/**
+* \brief           default CCM buffer encryption
+*
+* \param ctx       default CCM context
+* \param length    length of the input data in bytes
+* \param iv        nonce (initialization vector)
+* \param iv_len    length of IV in bytes
+*                  must be 2, 3, 4, 5, 6, 7 or 8
+* \param add       additional data
+* \param add_len   length of additional data in bytes
+*                  must be less than 2^16 - 2^8
+* \param input     buffer holding the input data
+* \param output    buffer for holding the output data
+*                  must be at least 'length' bytes wide
+* \param tag       buffer for holding the tag
+* \param tag_len   length of the tag to generate in bytes
+*                  must be 4, 6, 8, 10, 14 or 16
+*
+* \note            The tag is written to a separate buffer. To get the tag
+*                  concatenated with the output as in the CCM spec, use
+*                  tag = output + length and make sure the output buffer is
+*                  at least length + tag_len wide.
+*
+* \return          0 if successful
+*/
+int mbedtls_ccm_encrypt_and_tag_default( mbedtls_ccm_context_default *ctx, size_t length,
+                                         const unsigned char *iv, size_t iv_len,
+                                         const unsigned char *add, size_t add_len,
+                                         const unsigned char *input, unsigned char *output,
+                                         unsigned char *tag, size_t tag_len );
+
+/**
+* \brief           default CCM buffer authenticated decryption
+*
+* \param ctx       CCM default context
+* \param length    length of the input data
+* \param iv        initialization vector
+* \param iv_len    length of IV
+* \param add       additional data
+* \param add_len   length of additional data
+* \param input     buffer holding the input data
+* \param output    buffer for holding the output data
+* \param tag       buffer holding the tag
+* \param tag_len   length of the tag
+*
+* \return         0 if successful and authenticated,
+*                 MBEDTLS_ERR_CCM_AUTH_FAILED if tag does not match
+*/
```

Comment:
@gilles-peskine-arm @RonEld We should add here that implementations MUST NOT use the error code `MBEDTLS_ERR_CCM_AUTH_FAILED` for any kind of fatal internal error. This is because this error code is remapped to `MBEDTLS_ERR_CIPHER_AUTH_FAILED` which is remapped to `MBEDTLS_ERR_SSL_INVALID_MAC` which is IGNORED for DTLS. 

The fact that we have no clear separation of non-fatal and fatal errors is really unfortunate.